### DISCUSSION TITLE
Notification Plugin - How to

### DISCUSSION THREAD

**Discussion by Bbillyben** [inventree/InvenTree#4207] [Q&A]
Hi There, 

I would dev a plugin to [daily/weekly/userdefine] be notified of stock stale (and/or other stock management figures).

I would use the core notification process. and I'm currently into self feasibility assessment, so only ideas no script yet.

I'm trying to figure out how it's done for current event notification  (eg stock under minimum level, stock expiricy), but I can't find how a new message is added, how the target users are selected.

I've seen the comon.notification.NotificationMethod / BulkNotificationMethod where 'based' targets are defined in __init__
and the core plugin plugin.builtin.integration.core_notification.CoreNotificationsPlugin + EmailNotification
but I can't find where it's being instanciated (so how is defined the self.targets...)

Could you point me where to start digging into the notification process through inventree ?

**Comment by matmair** [comment]
Hi @Bbillyben!

`CoreNotificationMethod` etc. define the delivery methods that can be used by InvenTree. Every time `common.notifications.trigger_notification` gets called all notification methods (which are classes but are named NotificationMethod) are called in order of registration and get passed a target (user) list. This list is filtered (some notification methods might only delivery to a subset of users) and then the `send` method is called to initiate delivery.

Similar features to what you are proposing are already present so this should be doable.
There are a few steps to this, I will link some sample code:
1. Create a function in the plugin that gets scheduled - use the ScheduleMixin for this
https://github.com/inventree/InvenTree/blob/e730b5c24ce88b7b3d72d39903815359b5b66fdb/InvenTree/plugin/samples/integration/scheduled_task.py#L24-L62
3. The function should check for the conditions you use to determine if you should send a notification out (maybe make them concrete - one function per condition, one caller function that calls all the concrete ones)
https://github.com/inventree/InvenTree/blob/e730b5c24ce88b7b3d72d39903815359b5b66fdb/InvenTree/build/tasks.py#L37-L77
4. Determine who should get the notification
https://github.com/inventree/InvenTree/blob/e730b5c24ce88b7b3d72d39903815359b5b66fdb/InvenTree/part/models.py#L260-L278
5. Send the notification
https://github.com/inventree/InvenTree/blob/e730b5c24ce88b7b3d72d39903815359b5b66fdb/InvenTree/part/tasks.py#L19-L43

**Reply by Bbillyben** [reply]
Thanks @matmair  !

I can start digging the stream! 

I can see the instanciation of `NotificationMethod `in [def deliver_notification ](https://github.com/inventree/InvenTree/blob/e730b5c24ce88b7b3d72d39903815359b5b66fdb/InvenTree/common/notifications.py#L424) for instance.

**Reply by matmair** [reply]
@Bbillyben you can just ship a template with your plugin, should work out of the box
